src/System.CommandLine.Tests/Help/HelpBuilderTests.cs

```diff
@@ -1338,10 +1338,45 @@ public void Option_can_customize_descriptor()
             _console.ToString().Should().Contain(expected);
         }
 
+        [Fact]
+        public void Option_can_customize_descriptor_based_on_parse_result()
+        {
+            var option = new Option<bool>("option");
+            var commandA = new Command("a", $"a command help")
+            {
+                option
+            };
+            var commandB = new Command("b", $"b command help")
+            {
+                option
+            };
+            var command = new Command("root", "root command help")
+            {
+                commandA, commandB
+            };
+            var optionAHelpText = "option a help";
+            var optionBHelpText = "option b help";
+            Func<ParseResult, string> descriptionCallback = (ParseResult parseResult) =>
+                parseResult.CommandResult.Command.Equals(commandA) ? optionAHelpText : optionBHelpText;
+
+            HelpBuilder helpBuilder = new HelpBuilder(Resources.Instance, LargeMaxWidth);
+            helpBuilder.Customize(option, descriptor: descriptionCallback);
+
+            var parseResultA = command.Parse("root a --option");
+            helpBuilder.Write(parseResultA.CommandResult.Command, _console, parseResultA);
+            var help = _console.ToString();
```

Comment:
I would consider using two different console instances for the two different invocations since the content won't be cleared in between.